Ship Stargz Snapshotter

ktock opened the pull request:
#1981

This commit adds stargz snapshotter to kind.

Stargz snapshotter can be enabled by using `KIND_EXPERIMENTAL_CONTAINERD_SNAPSHOTTER=stargz` environment variable. This envvar is propagated to each node container and the entrypoint script modifies `/etc/containerd/config.toml` for stargz snapshotter and enables stargz snapshotter systemd service.

~~For enabling private registry authentication, stargz snapshotter requires kubeconfig that has `watch` and `list` access to secrets on the cluster's API server. So when `KIND_CONTAINERD_SNAPSHOTTER=stargz`, `kind create cluster` creates a service account that has these permissions and distributes the kubeconfig that contains the [service account token](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#service-account-tokens) to all nodes.~~ 
Since v0.6.1, additional kubeconfig is no longer needed for registry authentication because snapshotter aquires creds through CRI (https://github.com/containerd/stargz-snapshotter/pull/323).

## How to test

- base image built from this PR is available on: ghcr.io/ktock/kind-base:bda5c8b9
- node image built from this PR is available on: ghcr.io/ktock/kind-node:bda5c8b9

```
$ KIND_EXPERIMENTAL_CONTAINERD_SNAPSHOTTER=stargz kind create cluster --image ghcr.io/ktock/kind-node:bda5c8b9
```

Layers are mounted as FUSE instead of downloaded from the registry.

```
$ kubectl run test --restart=Never -it --image=ghcr.io/stargz-containers/python:3.7-esgz -- echo "hello"
$ docker exec kind-control-plane mount | grep "stargz on"
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/51/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/52/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/53/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/54/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/55/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/56/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/57/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/58/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/59/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
```

## Image size comparison

### additions by this PR

- base image
  - `containerd-stargz-grpc` (stargz snapshotter binary, v0.8.0) around 39 MB
  - fuse3 around 1 MB including deps (https://packages.ubuntu.com/focal/fuse3)
  - negligable static file changes
    - /etc/systemd/system/stargz-snapshotter.service
    - added few lines to /usr/local/bin/entrypoint
    - added few lines to /etc/containerd/config.toml

### size comparison with the `main` (f8e6aa66) and this PR (bda5c8b9) (around 39 MB addition)

**base image**

|version|image|size|
---|---|---
|main|ghcr.io/ktock/kind-base:f8e6aa66|around 279 MB
|this PR|ghcr.io/ktock/kind-base:bda5c8b9|around 318 MB|

**node image**

|version|image|size|
---|---|---
|main|ghcr.io/ktock/kind-node:f8e6aa66|around 1091 MB|
|this PR|ghcr.io/ktock/kind-node:bda5c8b9|around 1130 MB|

AkihiroSuda commented:
> Signed-off-by: ktock

Needs to be real name

 > commit 1d234a56

😆

AkihiroSuda commented on pkg/cluster/internal/create/actions/preparestargzsnapshotter/preparestargzsnapshotter.go:
```diff
@@ -0,0 +1,145 @@
+    certificate-authority-data: %s
+    server: https://%s:%d
+users:
+- name: snapshotter
```
The name should be "stargz-snapshotter"?

AkihiroSuda commented on pkg/cluster/internal/create/actions/preparestargzsnapshotter/preparestargzsnapshotter.go:
```diff
@@ -0,0 +1,145 @@
+clusters:
+- name: target-cluster
+  cluster:
+    certificate-authority-data: %s
```
Can we use text/template?

neolit123 commented on images/base/Dockerfile:
```diff
@@ -124,8 +131,17 @@ RUN echo "Ensuring scripts are executable ..." \
+    && rm -f /usr/local/bin/ctr-remote \
     && containerd --version \
     && runc --version \
+    && containerd-stargz-grpc --help \
```
would be nice to have --version added to this CLI eventually.

ktock replied:
Started to track this at https://github.com/containerd/stargz-snapshotter/issues/268.

neolit123 commented on pkg/build/nodeimage/const_stargz_snapshotter.go:
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
2021

neolit123 commented on pkg/cluster/internal/create/create.go:
```diff
@@ -122,6 +124,12 @@ func Cluster(logger log.Logger, p providers.Provider, opts *ClusterOptions) erro
 			waitforready.NewAction(opts.WaitForReady), // wait for cluster readiness
 		)
+		// prepare stargz snapshotter serviceaccount if needed
+		if v, ok := os.LookupEnv("KIND_EXPERIMENTAL_STARGZ"); ok && v == "1" {
```
are there plans to enable this snapshotter by default eventually?
if so maybe add a TODO + linking to an issue on top of this line.

neolit123 replied:
should this PR also touch the kind documentation?

neolit123 commented on pkg/cluster/internal/create/actions/preparestargzsnapshotter/preparestargzsnapshotter.go:
```diff
@@ -0,0 +1,145 @@
+	if err != nil {
+		return err
+	}
+	node := controlPlanes[0] // kind expects at least one always
```
i think kind had a method to get the "primary" control plane node?

neolit123 commented on pkg/build/nodeimage/const_stargz_snapshotter.go:
```diff
@@ -0,0 +1,52 @@
+rules:
+- apiGroups: [""]
+  resources: ["secrets"]
+  verbs: ["list", "watch"]
```
this grants all nodes the capability to list and watch secrets.
can you explain the requirements here - why does the snapshotter needs this?

neolit123 replied:
for the kind usage or anyone comfortable with the origin of the stargz binary (i.e. a trusted source code build), this is fine.
RBAC is limited here and does not allow limiting the client to only get the secret type corev1.SecretTypeDockerConfigJson. one has to add external admission for that.

BenTheElder replied:
For shipping to production clusters this seems a bit problematic though (adding another per node thing listing these) -- this seems like a flaw in containerd-cri + plugins.

BenTheElder replied:
This approach also won't work with the standard approach to giving kubelet on-disk credentials, which our users leverage.

neolit123 replied:
i agree, the way this is managed does feel like a workaround.

ktock replied:
We started to track this issue on https://github.com/containerd/stargz-snapshotter/issues/270.
How about limiting the lazy pulling functionality only to public repositores at this moment?

ktock replied:
This is fixed since stargz snapshotter v0.6.1. 
Since v0.6.1, additional kubeconfig is no longer needed for registry authentication because snapshotter acquires creds through CRI (containerd/stargz-snapshotter#323).

neolit123 commented on pkg/cluster/internal/create/actions/preparestargzsnapshotter/preparestargzsnapshotter.go:
```diff
@@ -0,0 +1,145 @@
+`, ca, server, common.APIServerInternalPort, string(tokenBytes))
+	for _, node := range allNodes {
+		if err := node.Command(
+			"/bin/sh", "-c", "touch /etc/kubernetes/stargz-snapshotter.conf && chmod 600 /etc/kubernetes/stargz-snapshotter.conf",
```
/etc/kubernetes/stargz-snapshotter.conf can be a constant scoped in this loop.

neolit123 commented on pkg/cluster/internal/create/actions/preparestargzsnapshotter/preparestargzsnapshotter.go:
```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
2021

ktock commented:
@AkihiroSuda @neolit123

Thanks for the comments.
Fixed except where I commented in the review threads.

BenTheElder commented on images/base/Dockerfile:
```diff
@@ -95,7 +102,7 @@ RUN echo "Ensuring scripts are executable ..." \
       libseccomp2 pigz \
       bash ca-certificates curl rsync \
-      nfs-common \
+      nfs-common fuse3 \
```
How large are these additions?

neolit123 replied:
looking at https://packages.debian.org/buster/fuse3 and dependencies, my estimate is around ~2.5MB, but would let @ktock confirm.

BenTheElder replied:
Er sorry: this addition but also the rest of them: how much larger is the image overall?
Will investigate in my next pass at this

ktock replied:
Added size comparison (**Image size comparison** section) to the above PR description.

[summary] around 33 MB addition by fuse3 + `containerd-stargz-grpc` binary (stargz snapshotter) + negligable config file changes

AkihiroSuda commented on images/base/files/usr/local/bin/entrypoint:
```diff
@@ -23,6 +23,19 @@ configure_containerd() {
   if [[ "$(stat -f -c %T /kind)" == 'zfs' ]]; then
     sed -i 's/snapshotter = "overlayfs"/snapshotter = "native"/' /etc/containerd/config.toml
   fi
+  if [[ "${KIND_EXPERIMENTAL_STARGZ:-}" == "1" ]] ; then
```
For extensibility, the env var name should be like KIND_CONTAINERD_SNAPSHOTTER=stargz

ktock replied:
Fixed it.

ktock commented:
Bumped stargz snapshotter to v0.6.1 and addressed review https://github.com/kubernetes-sigs/kind/pull/2076#pullrequestreview-591224262 .
Since v0.6.1, additional kubeconfig is no longer needed for registry authentication because snapshotter acquires creds through CRI (https://github.com/containerd/stargz-snapshotter/pull/323).

PTAL :pray:

BenTheElder commented:
I'd still have liked to see more discussion in the original issue covering e.g.:
- logistics of the binary, including keeping up with arbitrary containerd versions to pickup fixes and how this can be managed for those that don't use it but need to build from source for compliance
- the cost of adding this (image size, any changes we must make to bootstrapping and config)

I should have made that clearer. The issue title suggests we should just do it, but I mean for this part:
> Consider shipping stargz /crfs support in containerd by default for faster image pulls with images that support it.

To come through first before taking up work.

I'm a bit concerned that since filing the initial issue I've not seen much further interest expressed or uptake of the image format yet we're going to incur additional cost here managing this (and haven't really explored that).

It looks like this would now be possible to do fully by having some custom additions to a standard base image with no changes to the kind binary, I think we should perhaps start by offering this option and go from there.

BenTheElder commented on images/base/files/usr/local/bin/entrypoint:
```diff
@@ -109,6 +109,21 @@ configure_containerd() {
+      sed -i 's/disable_snapshot_annotations = true/disable_snapshot_annotations = false/' /etc/containerd/config.toml
+      sed -i -E 's|KUBELET_EXTRA_ARGS=(.*)|KUBELET_EXTRA_ARGS=\1 --image-service-endpoint=unix:///run/containerd-stargz-grpc/containerd-stargz-grpc.sock|' /etc/default/kubelet
+      echo "image-endpoint: unix:///run/containerd-stargz-grpc/containerd-stargz-grpc.sock" >> /etc/crictl.yaml
+      mkdir /etc/containerd-stargz-grpc
```
why wouldn't we do this statically, or at least before enabling the service?

BenTheElder commented on pkg/cluster/internal/providers/docker/provision.go:
```diff
@@ -203,6 +204,11 @@ func commonArgs(cluster string, cfg *config.Cluster, networkName string, nodeNam
 		args = append(args, "-e", fmt.Sprintf("%s=%s", key, val))
 	}
 
+	if v, ok := os.LookupEnv("KIND_CONTAINERD_SNAPSHOTTER"); ok {
```
I'm not sure this makes sense versus just having it in a version of the image. but if it does then the pattern is `KIND_EXPERIMENTAL_*` for enabling things that aren't stable yet.

ktock commented:
@BenTheElder Thank you for your review. Fixed the code based on the comments.

> logistics of the binary, including keeping up with arbitrary containerd versions to pickup fixes and how this can be managed for those that don't use it but need to build from source for compliance

Stargz snapshotter is nightly tested with the nightly version of containerd.
How about adding the logic to release nightly builds of stargz snapshotter in kind-ci?

> the cost of adding this (image size, any changes we must make to bootstrapping and config)

As mentioned in **Image size comparison** section in PR description, the following changes are needed to the base image as of now.

- containerd-stargz-grpc (stargz snapshotter binary, v0.6.1) around 49 MB
- fuse3 around 1 MB including deps (https://packages.ubuntu.com/focal/fuse3)
- small changes to static files
  - /etc/systemd/system/stargz-snapshotter.service : for managing stargz snapshotter service
  - adds few lines to /etc/containerd/config.toml  : for pluging stargz snapshotter to containerd
  - adds few lines to /usr/local/bin/entrypoint    : for enabling stargz snapshotter during startup the node

>  I'm a bit concerned that since filing the initial issue I've not seen much further interest expressed or uptake of the image format yet we're going to incur additional cost here managing this (and haven't really explored that).

Sorry for the long stop of this PR. We've been working on https://github.com/kubernetes-sigs/kind/pull/2076#pullrequestreview-591224262 and maturing snapshotter implementation. If you are interested in the recent adoption status of lazy pulling of stargz snapshotter (stargz/eStargz), please check out the recent blog https://medium.com/nttlabs/lazy-pulling-estargz-ef35812d73de

> It looks like this would now be possible to do fully by having some custom additions to a standard base image with no changes to the kind binary, I think we should perhaps start by offering this option and go from there.

So instead of customizing the default base image, we should have a new option to kind for building stargz-snapshotter-enabled node image and/or have an additional release of customized image tagged something like `kindest/node:vX.X.X-stargz-snapshotter`?

AkihiroSuda commented:
Needs rebase

ktock commented:
Rebased.
Can we move this forward?

aojea commented on images/base/Dockerfile:
```diff
@@ -196,6 +205,17 @@ RUN echo "Installing containerd-fuse-overlayfs ..." \
+    && curl -sSL --retry 5 --output /tmp/stargz-snapshotter.${TARGETARCH}.tgz "${STARGZ_SNAPSHOTTER_URL}" \
+    && echo "${STARGZ_SNAPSHOTTER_AMD64_SHA256SUM}  /tmp/stargz-snapshotter.amd64.tgz" | tee /tmp/stargz-snapshotter.sha256 \
+    && echo "${STARGZ_SNAPSHOTTER_ARM64_SHA256SUM}  /tmp/stargz-snapshotter.arm64.tgz" | tee -a /tmp/stargz-snapshotter.sha256 \
+    && echo "${STARGZ_SNAPSHOTTER_PPC64LE_SHA256SUM}  /tmp/stargz-snapshotter.ppc64le.tgz" | tee -a /tmp/stargz-snapshotter.sha256 \
```
are there S390X too?
we have them for others
```
    && echo "${CONTAINERD_FUSE_OVERLAYFS_S390X_SHA256SUM}  /tmp/containerd-fuse-overlayfs.s390x.tgz" | tee -a /tmp/containerd-fuse-overlayfs.sha256 \
```

ktock replied:
Thank you for the review. Added.

ktock replied:
Sorry, it was still missing. Added now.

koxu1996 commented:
Any ETA?
